transports/tls/tests/smoke.rs

```diff
@@ -0,0 +1,73 @@
+use futures::{future, StreamExt};
+use libp2p::multiaddr::Protocol;
+use libp2p::swarm::{DummyBehaviour, KeepAlive, SwarmEvent};
+use libp2p::Swarm;
+use libp2p_core::transport::MemoryTransport;
+use libp2p_core::upgrade::Version;
+use libp2p_core::Transport;
+
+#[tokio::test]
+async fn can_establish_connection() {
+    let mut swarm1 = make_swarm();
+    let mut swarm2 = make_swarm();
+
+    let listen_address = {
+        let expected_listener_id = swarm1.listen_on(Protocol::Memory(0).into()).unwrap();
+
+        loop {
+            match swarm1.next().await.unwrap() {
+                SwarmEvent::NewListenAddr {
+                    address,
+                    listener_id,
+                } if listener_id == expected_listener_id => break address,
+                _ => continue,
+            };
+        }
+    };
+    swarm2.dial(listen_address).unwrap();
```

Comment:
This one dials something like `/memory/2698462084451899396` (without PK). It would be nice to have two more tests: with a correct explicit PK and with incorrect PK.

Reply:
Verifying whether we dialed the correct peer in case the multiaddr has a peer ID inlined happens on the `Swarm` level in `rust-libp2p` and is thus transport independent. I don't think we need to add a test for this!

Reply:
That works, but it’s not a very nice behavior. Certificates should be verified during the handshake, as this allows you to abort the connection attempt during TLS handshake.
Verifying the peer ID after the handshake leads to the successful establishment of a connection on the server side, just to have that connection killed 1 RTT later. Even worse, while aborting during the handshake lets the peer know why the handshake failed (as this is communicated by a TLS alert), aborting after handshake completion leaves the peer guessing what went wrong.

Reply:
Yeah that makes sense. It is a historical decision that would require a fair bit of work to change. I'll open an issue about it but would prefer to not block this PR on it.

Reply:
https://github.com/libp2p/rust-libp2p/pull/2947